### PR TITLE
expression, executor: introduce propagateType for castDecimalAsReal

### LINKED ISSUE #26348: different result between mysql and tidb
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` sql
CREATE TABLE `t` (
`last_zccural_date` varchar(8) DEFAULT NULL,
`last_run_date` varchar(8) DEFAULT NULL,
`accr_amt` decimal(20,2) DEFAULT NULL,
`real_rate` decimal(15,8) DEFAULT NULL
);
insert into t values(20210606, 20210606, 50000.00, 5.04600000);
select last_zccural_date * accr_amt *(real_rate/36000) from t;
```

### 2. What did you expect to see? (Required)
``` sql
mysql> select last_zccural_date * accr_amt *(real_rate/36000) from t;
+-------------------------------------------------+
| last_zccural_date * accr_amt *(real_rate/36000) |
+-------------------------------------------------+
|                              141642663.71666598 |
+-------------------------------------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
``` sql
tidb> select last_zccural_date * accr_amt *(real_rate/36000) from t;
+-------------------------------------------------+
| last_zccural_date * accr_amt *(real_rate/36000) |
+-------------------------------------------------+
|                               141642664.0535101 |
+-------------------------------------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

014005ab68b5c9a28795d13c1ca685cf46a3be7e

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #26348 

Problem Summary:

### What is changed and how it works?

What's Changed:
Introduce the `PropagateType` interface for `Expression`. This interface
is used to propagate a specific type to an expression.

When building the `castDecimalAsReal`, we need to propagate the type of double into the argument.
Thus the argument can keep the precision as accurate as possible.

MySQL code reference:
Item_typecast_real::resolve_type
https://github.com/mysql/mysql-server/blob/beb865a960b9a8a16cf999c323e46c5b0c67f21f/sql/item_func.h#L1057 

Item_param::propagate_type
https://github.com/mysql/mysql-server/blob/beb865a960b9a8a16cf999c323e46c5b0c67f21f/sql/item.cc#L3622

set_data_type_double
https://github.com/mysql/mysql-server/blob/beb865a960b9a8a16cf999c323e46c5b0c67f21f/sql/item.h#L1365

float_length
https://github.com/mysql/mysql-server/blob/beb865a960b9a8a16cf999c323e46c5b0c67f21f/sql/item.h#L2130

DECIMAL_NOT_SPECIFIED
https://github.com/mysql/mysql-server/blob/beb865a960b9a8a16cf999c323e46c5b0c67f21f/storage/ndb/include/kernel/signaldata/DictTabInfo.hpp#L47

How it Works:

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Fix the issue that `cast(decimal_func as double)` is not compatible with MySQL.
```

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1011,6 +1011,38 @@ type builtinCastDecimalAsRealSig struct {
 	baseBuiltinCastFunc
 }
 
+func (b *builtinCastDecimalAsRealSig) PropagateType() {
```
Comment (wshwsh12) [inline expression/builtin_cast.go:1014]: Can we use normal `func ProgateRealType(arg Expression)` to do this? So we can call the function in other builtinFunction(eg. `builtinCastStringAsRealSig`) in future?

**Review by wshwsh12 — APPROVED** [review]

**Review by ichn-hu — APPROVED** [review]

**Pull request merged**